### PR TITLE
refactor(template-model): Get rid of BaseModel for BaseTemplate model

### PR DESCRIPTION
**Pull request by HeyRoach**

### CODE REVIEW

File: src/app/shared/services/tags/template-tag.service.ts [hunk 1]
```diff
@@ -62,4 +62,8 @@ export class TemplateTagService implements EntityTagService {
 
     return Observable.of(this.tagService.getValueFromTag(agreement) || null);
   }
+
+  public getResourceType(template) {
+    return template && template['bootable'] !== undefined ? 'Iso' : 'Template';
```
Comment (zolotyx) [inline src/app/shared/services/tags/template-tag.service.ts]: Please use proper typing  and get rid of square brakets

File: src/app/template/shared/base-template.model.ts [hunk 2]
```diff
@@ -1,74 +1,56 @@
...
-  public abstract get isTemplate(): boolean;
+export const sizeInGB = (template: BaseTemplateModel): number => {
+  return Utils.convertToGb(template.size);
+};
 
-  public get sizeInGB(): number {
-    return Utils.convertToGb(this.size);
-  }
+export const isTemplate = (template: BaseTemplateModel): boolean =>
+  template && template['bootable'] !== undefined ? false : true;
```
Comment (zolotyx) [inline src/app/template/shared/base-template.model.ts]: remove square brackets, use proper typing

File: src/app/template/shared/base-template.model.ts [hunk 3]
```diff
@@ -1,74 +1,56 @@
...
 
-  public get downloadUrl(): string {
-    const tag = this.tags.find(_ => _.key === TemplateTagKeys.downloadUrl);
+export const resourceType = (template: BaseTemplateModel) =>
```
Comment (zolotyx) [inline src/app/template/shared/base-template.model.ts]: please consider adding a TemplateType string enum and add a return type to this function

File: src/testutils/mocks/model-services/services/mock-iso.service.spec.ts [hunk 4]
```diff
@@ -8,6 +8,6 @@ const isos: Array<Object> = require('../fixtures/diskOfferings.json');
 @Injectable()
 export class MockIsoService {
   public getList(): Observable<Array<Iso>> {
-    return Observable.of(isos.map(json => new Iso(json)));
+    return Observable.of(isos.map((json: Iso) => json));
```
Comment (zolotyx) [inline src/testutils/mocks/model-services/services/mock-iso.service.spec.ts]: do you need a map?

File: src/testutils/mocks/model-services/services/mock-template.service.spec.ts [hunk 5]
```diff
@@ -11,7 +11,7 @@ export class MockTemplateService {
   constructor(@Inject('mockTemplateServiceConfig') public config: { templates: any, groupedTemplates: any }) {}
 
   public getList(): Observable<Array<Template>> {
-    return Observable.of(templates.map(json => new Template(json)));
+    return Observable.of(templates.map((json: Template) => json));
```
Comment (zolotyx) [inline src/testutils/mocks/model-services/services/mock-template.service.spec.ts]: remove map

File: src/app/template/shared/iso.model.ts [hunk 6]
```diff
@@ -2,15 +2,8 @@ import { BaseTemplateModel } from './base-template.model';
...
-  public path = 'iso';
-
-  public bootable: boolean;
-  public checksum: string;
-  public size: number;
-
-  public get isTemplate(): boolean {
-    return false;
-  }
+export interface Iso extends BaseTemplateModel, Taggable {
```
Comment (tamazlykar) [inline src/app/template/shared/iso.model.ts]: Please remove the Taggable interface. We need to explicitly write all properties. This interface will be removed in future.

File: src/app/template/shared/template.model.ts [hunk 7]
```diff
@@ -1,25 +1,12 @@
...
-  public hypervisor: string;
-  public isPasswordEnabled: boolean;
-  public status: string;
-  public type: string;
-  public size: number;
-
-  public get isTemplate(): boolean {
-    return true;
-  }
+export interface Template extends BaseTemplateModel, Taggable {
```
Comment (tamazlykar) [inline src/app/template/shared/template.model.ts]: Please remove the Taggable interface. We need to explicitly write all properties. This interface will be removed in future.

File: src/app/template/shared/template.service.spec.ts [hunk 8]
```diff
@@ -62,10 +63,10 @@ describe('Template service test', () => {
       groupId: 'group1',
       entity: 'Template'
     };
-    const template1 = new Template(params);
+    const template1 = params;
     const template2 = Object.assign(
       {},
```
Comment (tamazlykar) [inline src/app/template/shared/template.service.spec.ts]: use object spread

File: src/app/template/shared/base-template.model.ts [hunk 9]
```diff
@@ -1,74 +1,59 @@
...
-  public osTypeId: string;
-  public osTypeName: string;
-  public osType: OsType;
-  public size: number;
-  public status: string;
-  public tags: Array<Tag>;
-  public zoneId: string;
-  public zoneName: string;
-  public agreementAccepted?: boolean;
+export interface BaseTemplateModel extends BaseModelInterface, Taggable {
```
Comment (tamazlykar) [inline src/app/template/shared/base-template.model.ts]: Can you inspect taggable interface? Why it duplicate id property and allow to add any property to object? Can we remove these properties?

File: src/app/template/shared/template.model.ts (hunk 7 above)
Comment (tamazlykar) [inline src/app/template/shared/template.model.ts]: Taggable already extended on BaseTemplateInterface

File: src/app/template/template-list/template-card-list.component.ts [hunk 10]
```diff
@@ -45,7 +45,7 @@ export class TemplateCardListComponent {
   }
 
   public selectTemplate(template: BaseTemplateModel): void {
-    this.listService.showDetails(`${template.path}/${template.id}`);
+    this.listService.showDetails(`${getPath(template)}/${template.id}`); // getPath(template)
```
Comment (tamazlykar) [inline src/app/template/template-list/template-card-list.component.ts]: comment?

File: src/app/template/shared/base-template.model.ts [hunk 11]
```diff
@@ -1,74 +1,58 @@
...
-  public osTypeName: string;
-  public osType: OsType;
-  public size: number;
-  public status: string;
-  public tags: Array<Tag>;
-  public zoneId: string;
-  public zoneName: string;
-  public agreementAccepted?: boolean;
+export const resourceType = (template: BaseTemplateModel): TemplateResourceType =>
+  template && isTemplate(template) ? TemplateResourceType.Template : TemplateResourceType.Iso;
```
Comment (tamazlykar) [inline src/app/template/shared/base-template.model.ts]: template?

File: src/app/shared/services/tags/template-tag.service.ts [hunk 12]
```diff
@@ -62,4 +63,8 @@ export class TemplateTagService implements EntityTagService {
 
     return Observable.of(this.tagService.getValueFromTag(agreement) || null);
   }
+
+  public getResourceType(template: BaseTemplateModel): TemplateResourceType {
```
Comment (tamazlykar) [inline src/app/shared/services/tags/template-tag.service.ts]: why this duplicate code from base-template.model.ts?

File: src/app/template/shared/base-template.model.ts [hunk 13]
```diff
@@ -1,74 +1,58 @@
...
 
-  public zones?: Array<Partial<BaseTemplateModel>>;
+export const getPath = (template: BaseTemplateModel) =>
+  template && isTemplate(template) ? 'template' : 'iso';
```
Comment (tamazlykar) [inline src/app/template/shared/base-template.model.ts]: template?

**Pull request merged**